File: vm/minimal/Nga.hs
-- Minimal Nga virtual machine for RETRO, implemented in Haskell (GHC).
-- Builds to a single module: ghc Nga.hs -o nga-hs
import qualified Data.ByteString as BS
import Data.Array.IO (IOArray, newArray, readArray, writeArray)
import Data.Bits ((.&.), (.|.), shiftL, shiftR, complement, xor)
import Data.IORef
import Data.Int (Int32)
import Data.Word (Word32)
import Control.Monad (forM_, when)
import System.IO (hFlush, stdout, hIsEOF, stdin)
import System.Exit (exitSuccess)
import Data.Char (chr)

type Cell = Int32

imageSize, addresses, stackDepth :: Int
imageSize = 65536
addresses = 256
stackDepth = 256

cellMin, cellMax :: Cell
cellMin = minBound + 1
cellMax = maxBound - 1

data NgaVm = NgaVm
  { spRef     :: IORef Int
  , rpRef     :: IORef Int
  , ipRef     :: IORef Int
  , dataStack :: IOArray Int Cell
  , addrStack :: IOArray Int Cell
  , memory    :: IOArray Int Cell
  }

newVm :: IO NgaVm
newVm = do
  sp <- newIORef 0
  rp <- newIORef 0
  ip <- newIORef 0
  ds <- newArray (0, stackDepth - 1) 0
  as <- newArray (0, addresses - 1) 0
  mem <- newArray (0, imageSize) 0
  pure NgaVm { spRef = sp, rpRef = rp, ipRef = ip, dataStack = ds, addrStack = as, memory = mem }

tos :: NgaVm -> IO Cell
tos vm = do s <- readIORef (spRef vm); readArray (dataStack vm) s

nos :: NgaVm -> IO Cell
nos vm = do s <- readIORef (spRef vm); readArray (dataStack vm) (s - 1)

tors :: NgaVm -> IO Cell
tors vm = do r <- readIORef (rpRef vm); readArray (addrStack vm) r

setTos :: NgaVm -> Cell -> IO ()
setTos vm v = do s <- readIORef (spRef vm); writeArray (dataStack vm) s v

setNos :: NgaVm -> Cell -> IO ()
setNos vm v = do s <- readIORef (spRef vm); writeArray (dataStack vm) (s - 1) v

setTors :: NgaVm -> Cell -> IO ()
setTors vm v = do r <- readIORef (rpRef vm); writeArray (addrStack vm) r v

stackPush :: NgaVm -> Cell -> IO ()
stackPush vm v = do
  modifyIORef' (spRef vm) (+1)
  s <- readIORef (spRef vm)
  writeArray (dataStack vm) s v

stackPop :: NgaVm -> IO Cell
stackPop vm = do
  s <- readIORef (spRef vm)
  v <- readArray (dataStack vm) s
  modifyIORef' (spRef vm) (subtract 1)
  pure v

execute :: NgaVm -> Int -> IO ()
execute vm cell = do
  writeIORef (rpRef vm) 1
  writeIORef (ipRef vm) cell
  loop
 where
  loop = do
    ip <- readIORef (ipRef vm)
    when (ip < imageSize) $ do
      opcode <- readArray (memory vm) ip
      processOpcodeBundle vm opcode
      modifyIORef' (ipRef vm) (+1)
      rp <- readIORef (rpRef vm)
      when (rp == 0) $ writeIORef (ipRef vm) imageSize
      loop

loadImage :: NgaVm -> FilePath -> IO ()
loadImage vm path = do
  bytes <- BS.readFile path
  let cells = min imageSize (BS.length bytes `div` 4)
  forM_ [0 .. cells - 1] $ \i -> do
    let idx = i * 4
        b0 = fromIntegral (BS.index bytes idx)       :: Word32
        b1 = fromIntegral (BS.index bytes (idx + 1)) :: Word32
        b2 = fromIntegral (BS.index bytes (idx + 2)) :: Word32
        b3 = fromIntegral (BS.index bytes (idx + 3)) :: Word32
        w  = b0 .|. (b1 `shiftL` 8) .|. (b2 `shiftL` 16) .|. (b3 `shiftL` 24)
        c  = fromIntegral w :: Cell
    writeArray (memory vm) i c

prepareVm :: NgaVm -> IO ()
prepareVm vm = do
  writeIORef (ipRef vm) 0
  writeIORef (spRef vm) 0
  writeIORef (rpRef vm) 0
  forM_ [0 .. imageSize] $ \i -> writeArray (memory vm) i 0
  forM_ [0 .. stackDepth - 1] $ \i -> writeArray (dataStack vm) i 0
  forM_ [0 .. addresses - 1] $ \i -> writeArray (addrStack vm) i 0

inst_no :: NgaVm -> IO ()
inst_no _ = pure ()

inst_li :: NgaVm -> IO ()
inst_li vm = do
  modifyIORef' (ipRef vm) (+1)
  ip <- readIORef (ipRef vm)
  v <- readArray (memory vm) ip
  stackPush vm v

inst_du :: NgaVm -> IO ()
inst_du vm = tos vm >>= stackPush vm

inst_dr :: NgaVm -> IO ()
inst_dr vm = do
  s <- readIORef (spRef vm)
  writeArray (dataStack vm) s 0
  modifyIORef' (spRef vm) (subtract 1)
  s' <- readIORef (spRef vm)
  when (s' < 0) $ writeIORef (ipRef vm) imageSize

inst_sw :: NgaVm -> IO ()
inst_sw vm = do
  a <- tos vm
  b <- nos vm
  setTos vm b
  setNos vm a

inst_pu :: NgaVm -> IO ()
inst_pu vm = do
  modifyIORef' (rpRef vm) (+1)
  t <- tos vm
  setTors vm t
  inst_dr vm

inst_po :: NgaVm -> IO ()
inst_po vm = do
  t <- tors vm
  stackPush vm t
  modifyIORef' (rpRef vm) (subtract 1)

inst_ju :: NgaVm -> IO ()
inst_ju vm = do
  t <- tos vm
  writeIORef (ipRef vm) (fromIntegral t - 1)
  inst_dr vm

inst_ca :: NgaVm -> IO ()
inst_ca vm = do
  modifyIORef' (rpRef vm) (+1)
  ip <- readIORef (ipRef vm)
  setTors vm (fromIntegral ip)
  t <- tos vm
  writeIORef (ipRef vm) (fromIntegral t - 1)
  inst_dr vm

inst_cc :: NgaVm -> IO ()
inst_cc vm = do
  a <- tos vm
  inst_dr vm
  b <- tos vm
  inst_dr vm
  when (b /= 0) $ do
    modifyIORef' (rpRef vm) (+1)
    ip <- readIORef (ipRef vm)
    setTors vm (fromIntegral ip)
    writeIORef (ipRef vm) (fromIntegral a - 1)

inst_re :: NgaVm -> IO ()
inst_re vm = do
  t <- tors vm
  writeIORef (ipRef vm) (fromIntegral t)
  modifyIORef' (rpRef vm) (subtract 1)

inst_eq :: NgaVm -> IO ()
inst_eq vm = do
  a <- nos vm
  b <- tos vm
  setNos vm (if a == b then -1 else 0)
  inst_dr vm

inst_ne :: NgaVm -> IO ()
inst_ne vm = do
  a <- nos vm
  b <- tos vm
  setNos vm (if a /= b then -1 else 0)
  inst_dr vm

inst_lt :: NgaVm -> IO ()
inst_lt vm = do
  a <- nos vm
  b <- tos vm
  setNos vm (if a < b then -1 else 0)
  inst_dr vm

inst_gt :: NgaVm -> IO ()
inst_gt vm = do
  a <- nos vm
  b <- tos vm
  setNos vm (if a > b then -1 else 0)
  inst_dr vm

inst_fe :: NgaVm -> IO ()
inst_fe vm = do
  t <- tos vm
  case t of
    -1 -> setTos vm . fromIntegral =<< fmap (subtract 1) (readIORef (spRef vm))
    -2 -> setTos vm . fromIntegral =<< readIORef (rpRef vm)
    -3 -> setTos vm (fromIntegral imageSize)
    -4 -> setTos vm cellMin
    -5 -> setTos vm cellMax
    _  -> do v <- readArray (memory vm) (fromIntegral t); setTos vm v

inst_st :: NgaVm -> IO ()
inst_st vm = do
  addr <- tos vm
  if addr >= 0 && addr <= fromIntegral imageSize
     then do
       val <- nos vm
       writeArray (memory vm) (fromIntegral addr) val
       inst_dr vm
       inst_dr vm
     else writeIORef (ipRef vm) imageSize

inst_ad :: NgaVm -> IO ()
inst_ad vm = do a <- nos vm; b <- tos vm; setNos vm (a + b); inst_dr vm

inst_su :: NgaVm -> IO ()
inst_su vm = do a <- nos vm; b <- tos vm; setNos vm (a - b); inst_dr vm

inst_mu :: NgaVm -> IO ()
inst_mu vm = do a <- nos vm; b <- tos vm; setNos vm (a * b); inst_dr vm

inst_di :: NgaVm -> IO ()
inst_di vm = do
  a <- tos vm
  b <- nos vm
  setTos vm (b `div` a)
  setNos vm (b `mod` a)

inst_an :: NgaVm -> IO ()
inst_an vm = do a <- tos vm; b <- nos vm; setNos vm (a .&. b); inst_dr vm

inst_or :: NgaVm -> IO ()
inst_or vm = do a <- tos vm; b <- nos vm; setNos vm (a .|. b); inst_dr vm

inst_xo :: NgaVm -> IO ()
inst_xo vm = do a <- tos vm; b <- nos vm; setNos vm (a `xor` b); inst_dr vm

inst_sh :: NgaVm -> IO ()
inst_sh vm = do
  y <- tos vm
  x <- nos vm
  let result
        | y < 0     = x `shiftL` fromIntegral (0 - y)
        | x < 0 && y > 0 = (x `shiftR` fromIntegral y) .|. (complement 0 `shiftR` fromIntegral y)
        | otherwise = x `shiftR` fromIntegral y
  setNos vm result
  inst_dr vm

inst_zr :: NgaVm -> IO ()
inst_zr vm = do
  t <- tos vm
  when (t == 0) $ do
    inst_dr vm
    r <- tors vm
    writeIORef (ipRef vm) (fromIntegral r)
    modifyIORef' (rpRef vm) (subtract 1)

inst_ha :: NgaVm -> IO ()
inst_ha vm = writeIORef (ipRef vm) imageSize

inst_ie :: NgaVm -> IO ()
inst_ie vm = stackPush vm 2

inst_iq :: NgaVm -> IO ()
inst_iq vm = do
  t <- tos vm
  case t of
    0 -> do inst_dr vm; stackPush vm 0; stackPush vm 0
    1 -> do inst_dr vm; stackPush vm 1; stackPush vm 1
    _ -> pure ()

inst_ii :: NgaVm -> IO ()
inst_ii vm = do
  t <- tos vm
  case t of
    0 -> do
      inst_dr vm
      c <- stackPop vm
      putChar (chr (fromIntegral (c .&. 0xFF)))
      hFlush stdout
    1 -> do
      eof <- hIsEOF stdin
      when eof exitSuccess
      b <- getChar
      inst_dr vm
      stackPush vm (fromIntegral (fromEnum b .&. 0xFF))
    _ -> inst_dr vm

processOpcodeBundle :: NgaVm -> Cell -> IO ()
processOpcodeBundle vm opcode = do
  executeInstruction vm (fromIntegral (opcode .&. 0xFF))
  executeInstruction vm (fromIntegral ((opcode `shiftR` 8) .&. 0xFF))
  executeInstruction vm (fromIntegral ((opcode `shiftR` 16) .&. 0xFF))
  executeInstruction vm (fromIntegral ((opcode `shiftR` 24) .&. 0xFF))

executeInstruction :: NgaVm -> Int -> IO ()
executeInstruction vm inst = case inst of
  0  -> inst_no vm
  1  -> inst_li vm
  2  -> inst_du vm
  3  -> inst_dr vm
  4  -> inst_sw vm
  5  -> inst_pu vm
  6  -> inst_po vm
  7  -> inst_ju vm
  8  -> inst_ca vm
  9  -> inst_cc vm
  10 -> inst_re vm
  11 -> inst_eq vm
  12 -> inst_ne vm
  13 -> inst_lt vm
  14 -> inst_gt vm
  15 -> inst_fe vm
  16 -> inst_st vm
  17 -> inst_ad vm
  18 -> inst_su vm
  19 -> inst_mu vm
  20 -> inst_di vm
  21 -> inst_an vm
  22 -> inst_or vm
  23 -> inst_xo vm
  24 -> inst_sh vm
  25 -> inst_zr vm
  26 -> inst_ha vm
  27 -> inst_ie vm
  28 -> inst_iq vm
  29 -> inst_ii vm
  _  -> inst_no vm

main :: IO ()
main = do
  vm <- newVm
  prepareVm vm
  loadImage vm "ngaImage"
  execute vm 0

File: vm/minimal/nga.js
// Minimal Nga virtual machine for RETRO, implemented for Node.js.
// Mirrors the reference VM in nga.c/nga.rs.
const fs = require('fs');

const IMAGE_SIZE = 65536;
const ADDRESSES = 256;
const STACK_DEPTH = 256;

const CELL_MIN = -0x7fffffff - 1 + 1; // INT_MIN + 1
const CELL_MAX = 0x7fffffff - 1;      // INT_MAX - 1

const toCell = (n) => (n | 0);

class NgaVm {
  constructor() {
    this.sp = 0;
    this.rp = 0;
    this.ip = 0;
    this.data = new Int32Array(STACK_DEPTH);
    this.address = new Int32Array(ADDRESSES);
    this.memory = new Int32Array(IMAGE_SIZE + 1);
  }

  halt() { this.ip = IMAGE_SIZE; }
  validAddress(addr) { return Number.isInteger(addr) && addr >= 0 && addr <= IMAGE_SIZE; }

  tos() {
    return this.data[this.sp];
  }

  nos() {
    return this.data[this.sp - 1];
  }

  tors() {
    return this.address[this.rp];
  }

  setTos(v) {
    this.data[this.sp] = toCell(v);
  }

  setNos(v) {
    this.data[this.sp - 1] = toCell(v);
  }

  setTors(v) {
    this.address[this.rp] = toCell(v);
  }

  stackPop() {
    if (this.sp <= 0) { this.halt(); return 0; }
    this.sp -= 1;
    return this.data[this.sp + 1];
  }

  stackPush(v) {
    if (this.sp >= STACK_DEPTH - 1) { this.halt(); return; }
    this.sp += 1;
    this.data[this.sp] = toCell(v);
  }

  execute(cell) {
    this.rp = 1;
    this.ip = cell | 0;
    while (this.ip >= 0 && this.ip < IMAGE_SIZE) {
      const opcode = this.memory[this.ip];
      this.processOpcodeBundle(opcode);
      this.ip += 1;
      if (this.rp === 0) {
        this.ip = IMAGE_SIZE;
      }
    }
  }

  loadImage(imageFile) {
    if (fs.statSync(imageFile).size > IMAGE_SIZE * 4) {
      throw new Error('image exceeds VM memory capacity');
    }
    const buf = fs.readFileSync(imageFile);
    const cells = Math.floor(buf.length / 4);
    for (let i = 0; i < cells && i < IMAGE_SIZE; i += 1) {
      this.memory[i] = buf.readInt32LE(i * 4);
    }
  }

  prepareVm() {
    this.ip = 0;
    this.sp = 0;
    this.rp = 0;
    this.memory.fill(0);
    this.data.fill(0);
    this.address.fill(0);
  }

  inst_no() {}

  inst_li() {
    if (this.ip >= IMAGE_SIZE - 1) { this.halt(); return; }
    this.ip += 1;
    this.stackPush(this.memory[this.ip]);
  }

  inst_du() {
    if (this.sp <= 0) { this.halt(); return; }
    this.stackPush(this.tos());
  }

  inst_dr() {
    this.data[this.sp] = 0;
    this.sp -= 1;
    if (this.sp < 0) {
      this.ip = IMAGE_SIZE;
    }
  }

  inst_sw() {
    if (this.sp < 2) { this.halt(); return; }
    const a = this.tos();
    this.setTos(this.nos());
    this.setNos(a);
  }

  inst_pu() {
    if (this.sp <= 0 || this.rp >= ADDRESSES - 1) { this.halt(); return; }
    this.rp += 1;
    this.setTors(this.tos());
    this.inst_dr();
  }

  inst_po() {
    if (this.rp <= 0) { this.halt(); return; }
    this.stackPush(this.tors());
    this.rp -= 1;
  }

  inst_ju() {
    if (this.sp <= 0 || !this.validAddress(this.tos()) || this.tos() === IMAGE_SIZE) { this.halt(); return; }
    this.ip = this.tos() - 1;
    this.inst_dr();
  }

  inst_ca() {
    if (this.sp <= 0 || this.rp >= ADDRESSES - 1 || !this.validAddress(this.tos()) || this.tos() === IMAGE_SIZE) { this.halt(); return; }
    this.rp += 1;
    this.setTors(this.ip);
    this.ip = this.tos() - 1;
    this.inst_dr();
  }

  inst_cc() {
    if (this.sp < 2) { this.halt(); return; }
    const a = this.tos();
    this.inst_dr();
    const b = this.tos();
    this.inst_dr();
    if (b !== 0) {
      if (this.rp >= ADDRESSES - 1 || !this.validAddress(a) || a === IMAGE_SIZE) { this.halt(); return; }
      this.rp += 1;
      this.setTors(this.ip);
      this.ip = a - 1;
    }
  }

  inst_re() {
    if (this.rp <= 0 || !this.validAddress(this.tors()) || this.tors() === IMAGE_SIZE) { this.halt(); return; }
    this.ip = this.tors();
    this.rp -= 1;
  }

  inst_eq() {
    this.setNos(this.nos() === this.tos() ? -1 : 0);
    this.inst_dr();
  }

  inst_ne() {
    this.setNos(this.nos() !== this.tos() ? -1 : 0);
    this.inst_dr();
  }

  inst_lt() {
    this.setNos(this.nos() < this.tos() ? -1 : 0);
    this.inst_dr();
  }

  inst_gt() {
    this.setNos(this.nos() > this.tos() ? -1 : 0);
    this.inst_dr();
  }

  inst_fe() {
    if (this.sp <= 0) { this.halt(); return; }
    switch (this.tos()) {
      case -1:
        this.setTos(this.sp - 1);
        break;
      case -2:
        this.setTos(this.rp);
        break;
      case -3:
        this.setTos(IMAGE_SIZE);
        break;
      case -4:
        this.setTos(CELL_MIN);
        break;
      case -5:
        this.setTos(CELL_MAX);
        break;
      default:
        if (!this.validAddress(this.tos())) { this.halt(); return; }
        this.setTos(this.memory[this.tos()]);
        break;
    }
  }

  inst_st() {
    const addr = this.tos();
    if (this.sp < 2) {
      this.halt();
    } else if (this.validAddress(addr)) {
      this.memory[addr] = this.nos();
      this.inst_dr();
      this.inst_dr();
    } else {
      this.ip = IMAGE_SIZE;
    }
  }

  inst_ad() {
    this.setNos(toCell(this.nos() + this.tos()));
    this.inst_dr();
  }

  inst_su() {
    this.setNos(toCell(this.nos() - this.tos()));
    this.inst_dr();
  }

  inst_mu() {
    this.setNos(toCell(this.nos() * this.tos()));
    this.inst_dr();
  }

  inst_di() {
    if (this.sp < 2 || this.tos() === 0 || (this.nos() === CELL_MIN && this.tos() === -1)) { this.halt(); return; }
    const a = this.tos();
    const b = this.nos();
    this.setTos(toCell(b / a));
    this.setNos(toCell(b % a));
  }

  inst_an() {
    this.setNos(this.tos() & this.nos());
    this.inst_dr();
  }

  inst_or() {
    this.setNos(this.tos() | this.nos());
    this.inst_dr();
  }

  inst_xo() {
    this.setNos(this.tos() ^ this.nos());
    this.inst_dr();
  }

  inst_sh() {
    const y = this.tos();
    const x = this.nos();
    let result;
    if (this.sp < 2 || y < -31 || y > 31) { this.halt(); return; }
    if (y < 0) {
      result = x << (0 - y);
    } else if (x < 0 && y > 0) {
      result = (x >> y) | (~((~0) >>> y));
    } else {
      result = x >> y;
    }
    this.setNos(result);
    this.inst_dr();
  }

  inst_zr() {
    if (this.sp <= 0) { this.halt(); return; }
    if (this.tos() === 0) {
      if (this.rp <= 0 || !this.validAddress(this.tors()) || this.tors() === IMAGE_SIZE) { this.halt(); return; }
      this.inst_dr();
      this.ip = this.tors();
      this.rp -= 1;
    }
  }

  inst_ha() {
    this.ip = IMAGE_SIZE;
  }

  inst_ie() {
    this.stackPush(2);
  }

  inst_iq() {
    if (this.sp <= 0) { this.halt(); return; }
    if (this.tos() === 0) {
      this.inst_dr();
      this.stackPush(0);
      this.stackPush(0);
    } else if (this.tos() === 1) {
      this.inst_dr();
      this.stackPush(1);
      this.stackPush(1);
    }
  }

  inst_ii() {
    if (this.sp <= 0) { this.halt(); return; }
    if (this.tos() === 0) {
      this.inst_dr();
      const c = this.stackPop();
      process.stdout.write(String.fromCharCode(c & 0xff));
    } else if (this.tos() === 1) {
      const buf = Buffer.alloc(1);
      const read = fs.readSync(0, buf, 0, 1, null);
      if (read <= 0) {
        process.exit(0);
      }
      this.inst_dr();
      this.stackPush(buf[0]);
    } else {
      this.inst_dr();
    }
  }

  processOpcodeBundle(opcode) {
    if ((opcode & 0xff) > 29 || ((opcode >> 8) & 0xff) > 29 ||
        ((opcode >> 16) & 0xff) > 29 || ((opcode >> 24) & 0xff) > 29) {
      this.halt();
      return;
    }
    this.executeInstruction(opcode & 0xff);
    this.executeInstruction((opcode >> 8) & 0xff);
    this.executeInstruction((opcode >> 16) & 0xff);
    this.executeInstruction((opcode >> 24) & 0xff);
  }

  executeInstruction(inst) {
    switch (inst) {
      case 0: this.inst_no(); break;
      case 1: this.inst_li(); break;
      case 2: this.inst_du(); break;
      case 3: this.inst_dr(); break;
      case 4: this.inst_sw(); break;
      case 5: this.inst_pu(); break;
      case 6: this.inst_po(); break;
      case 7: this.inst_ju(); break;
      case 8: this.inst_ca(); break;
      case 9: this.inst_cc(); break;
      case 10: this.inst_re(); break;
      case 11: this.inst_eq(); break;
      case 12: this.inst_ne(); break;
      case 13: this.inst_lt(); break;
      case 14: this.inst_gt(); break;
      case 15: this.inst_fe(); break;
      case 16: this.inst_st(); break;
      case 17: this.inst_ad(); break;
      case 18: this.inst_su(); break;
      case 19: this.inst_mu(); break;
      case 20: this.inst_di(); break;
      case 21: this.inst_an(); break;
      case 22: this.inst_or(); break;
      case 23: this.inst_xo(); break;
      case 24: this.inst_sh(); break;
      case 25: this.inst_zr(); break;
      case 26: this.inst_ha(); break;
      case 27: this.inst_ie(); break;
      case 28: this.inst_iq(); break;
      case 29: this.inst_ii(); break;
      default: this.inst_no(); break;
    }
  }
}

function main() {
  const vm = new NgaVm();
  vm.prepareVm();
  vm.loadImage('ngaImage');
  vm.execute(0);
}

if (require.main === module) {
  main();
}
